[Merged by Bors] - feat(algebra/direct_sum): Add ⨁ notation

eric-wieser opened the pull request:
This uses the unicode character "n-ary circled plus operator", which seems to be the usual symbol for this operation


---
<!-- put comments you want to keep out of the PR commit here -->

bryangingechen reviewed (approved):
LGTM. Anyone else have comments?

eric-wieser commented on src/algebra/direct_sum.lean:
```diff
@@ -15,20 +25,22 @@ variables (ι : Type v) [decidable_eq ι] (β : ι → Type w) [Π i, add_comm_g
 
 def direct_sum : Type* := Π₀ i, β i
 
+notation `⨁` binders `, ` r:(scoped f, direct_sum _ f) := r
```
Should this be within the big operators locale?

eric-wieser commented on src/algebra/direct_sum.lean:
```diff
@@ -15,20 +25,22 @@ variables (ι : Type v) [decidable_eq ι] (β : ι → Type w) [Π i, add_comm_g
 
+notation `⨁` binders `, ` r:(scoped f, direct_sum _ f) := r
+
 namespace direct_sum
```
This should do the job, right?
```suggestion
namespace direct_sum

localized notation `⨁` binders `, ` r:(scoped f, direct_sum _ f) := r
```

bryangingechen replied:
This doesn't look like the right syntax. It should be ``localized notation `⨁` binders `, ` r:(scoped f, direct_sum _ f) := r in direct_sum``.

eric-wieser replied:
Does it still need to be within the namespace?

bryangingechen replied:
Oh. I got it wrong too. `notation ... := ...` needs to be in double-quotes: ``localized "notation `⨁` binders `, ` r:(scoped f, direct_sum _ f) := r" in direct_sum``.

bryangingechen replied:
And it doesn't have to be within the namespace.

eric-wieser commented on src/algebra/direct_sum.lean:
```diff
@@ -17,18 +27,20 @@ def direct_sum : Type* := Π₀ i, β i
 
 namespace direct_sum
 
+localized notation `⨁` binders `, ` r:(scoped f, direct_sum _ f) := r
```
```suggestion
localized notation `⨁` binders `, ` r:(scoped f, direct_sum _ f) := r in direct_sum
```

eric-wieser commented on src/linear_algebra/tensor_product.lean:
```diff
@@ -613,10 +613,11 @@ variables (M₁ : ι₁ → Type*) (M₂ : ι₂ → Type*)
 "tensor product distributes over direct sum". -/
-def direct_sum : direct_sum ι₁ M₁ ⊗[R] direct_sum ι₂ M₂
-  ≃ₗ[R] direct_sum (ι₁ × ι₂) (λ i, M₁ i.1 ⊗[R] M₂ i.2) :=
+def direct_sum : (⨁ i₁, M₁ i₁) ⊗[R] (⨁ i₂, M₂ i₂) ≃ₗ[R] (⨁ (i : ι₁ × ι₂), M₁ i.1 ⊗[R] M₂ i.2) :=
```
Hey look, we already used (almost) this notation in a comment!

Note that `⨁` (n-ary oplus) is the character my patch uses, and `⨁` (oplus) is what the original comment used.

eric-wieser committed:
feat(algebra/direct_sum): Add ⨁ notation

This uses the unicode character "n-ary circled plus operator", which seems to be the usual symbol for this operation

robertylewis commented on src/algebra/direct_sum.lean:
```diff
@@ -3,10 +3,20 @@ Copyright (c) 2019 Kenny Lau. All rights reserved.
+# Direct sum
+
+Direct sum of abelian groups, indexed by a discrete type.
+`⨁ i, β i` is the n-ary direct sum, accessible after `open_locale direct_sum`.
```
```suggestion
This file defines the direct sum of abelian groups, indexed by a discrete type.

## Notation

`⨁ i, β i` is the n-ary direct sum `direct_sum`.
This notation is in the `direct_sum` locale, accessible after `open_locale direct_sum`.
```

I wish this module doc had more detail, but that's not your problem in this PR.

robertylewis committed:
Update src/algebra/direct_sum.lean
